src/ui/Editor.h

```diff
@@ -1,225 +1,97 @@
-			void				ShowWhiteSpaces(bool show);
-			bool				LineEndingsVisible();
-			bool				WhiteSpacesVisible();
-
-
-			void				SetProjectFolder(ProjectFolder*);
-			ProjectFolder*		GetProjectFolder() const { return fProjectFolder; }
-			void				SetZoom(int32 zoom);
-			void				Undo();
-			void				Completion();
-			void				Format();
-			void				GoToDefinition();
-			void				GoToDeclaration();
-			void				GoToImplementation();
-			void				SwitchSourceHeader();
-			void				UncommentSelection();
-
-			void 				ContextMenu(BPoint point);
-
-			void				SetProblems(const BMessage* diagnostics);
-			void				GetProblems(BMessage* diagnostics);
-
-			filter_result		BeforeKeyDown(BMessage*);
-			filter_result		OnArrowKey(int8 ch);
-
-			std::string			FileType() const { return fFileType; }
-			void				SetFileType(std::string fileType) { fFileType = fileType; }
-
-			void				SetCommentLineToken(std::string commenter){ fCommenter = commenter; }
-			void				SetCommentBlockTokens(std::string startBlock, std::string endBlock){ /*TODO! */}
-
-			LSPEditorWrapper*	GetLSPEditorWrapper() { return fLSPEditorWrapper; }
-			bool				HasLSPServer() const;
-			bool				HasLSPCapability(const LSPCapability cap);
-
-
-private:
-			BString	const		_EndOfLineString();
-			void				UpdateStatusBar();
-			void				_ApplyExtensionSettings();
-			void				_MaintainIndentation(char c);
-			void				_SetLineIndentation(int line, int indent);
-			void				_BraceHighlight();
-			bool				_BraceMatch(int pos);
-			void				_CommentLine(int32 position);
-			void				_EndOfLineAssign(char *buffer, int32 size);
-			void				_HighlightBraces();
-			void				_RedrawNumberMargin(bool forced = false);
-			void				_SetFoldMargin(bool enabled);
-			void				_UpdateSavePoint(bool modified);
-
-			template<typename T>
-			typename T::type	Get() { return T::Get(this); }
-			template<typename T>
-			void				Set(typename T::type value) { T::Set(this, value); }
-
-private:
-
-			entry_ref			fFileRef;
-			bool				fModified;
-			BString				fFileName;
-			node_ref			fNodeRef;
-			BMessenger			fTarget;
-
-			bool				fBracingAvailable;
-			std::string			fFileType;
-			bool				fFoldingAvailable;
-			bool				fSyntaxAvailable;
-			bool				fParsingAvailable;
-			std::string			fCommenter;
-			int					fLinesLog10;
-
-			int					fCurrentLine;
-			int					fCurrentColumn;
-
-			LSPEditorWrapper*		fLSPEditorWrapper;
-			ProjectFolder*		fProjectFolder;
-			editor::StatusView*			fStatusView;
-
-			BMessage	fProblems;
-			BLocker		fProblemsLock;
+			virtual void	ApplySettings() = 0;
+
+			virtual bool	CanCopy() = 0;
+			virtual bool	CanCut() = 0;
+			virtual bool	CanPaste() = 0;
+			virtual bool	CanRedo() = 0;
+			virtual bool	CanUndo() = 0;
+			virtual void	Copy() = 0;
+			virtual void	Cut() = 0;
+			virtual void	Paste() = 0;
+			virtual void	Redo() = 0;
+			virtual void	Undo() = 0;
+
+			virtual	const 	BString		Selection() = 0;
 
-};
 
-#endif // EDITOR_H
+			//These are specific to CodeEditor
```

Comment:
That doesn't look correct to me, It would be more appropriate to include these methods in a separate CodeEditor interface which inherits from Editor.
If one wants to create a new Editor (let's say a BMessage editor) they would carry all the methods from CodeEditor that would not necessarily apply the new one.

Reply:
The comment is a bit misleading, it refers to the GoToLine and GoToLSPPosition only.

Anyway I agree that we could have a "in the middle" interface and I don't think we are preventing it.. I mean, as soon as we need to have it we could introduce it. 
Behind able to to the perfect 'cut' now is quite complex and mostly based on unknown future needs.
I would like the Editor to be as much clean as possible also to force us in using messages to share information between the editor and the window and decouple the two objects.. 
At some point we could reach an "Editor" interface (maybe at that point the name will need to change) also as a base class for the ProjectBrowser.. at that point I'm sure we will have a layered set of interfaces..